2023/incoming/the-magical-test2-suite-part-2.pod

```diff
@@ -0,0 +1,182 @@
+
+    my $obj = mock {
+        happy => mock {
+            holidays => mock {
+                everyone => 'Ho! Ho! Ho!'
+            }
+        }
+    };
+
+    is $obj->happy->holidays->everyone, 'Ho! Ho! Ho!', 'chained mocks';
+
+I<“But Santa”>, you may ask, I<< “What about when the class is instantiated and used somewhere inside the code I want to test? How can I use mocks to add or override methods to something out of my hands?” >> No worries! Just expand your mock definition for more control:
+
+    my $mock_meta = mock 'Some::Class' => (
+        track => true,
+        override => [
+            some_method  => sub { … },
+            other_method => sub { … },
+        ],
+        add => [
+            new_method => sub { … },
+        ],
+    );
+
+And now, as long as C<$mock_meta> exists, any instances of “C<Some::Class>” will have the mocked behavior. To restore them back to the original, simply undefine that C<$mock_meta> variable or call C<< $mock_meta->reset_all() >>.
+
+One thing you may have noticed in the example above is the “C<< track => true >>”. That means C<$mock_meta> will contain a lot of information ready for you to inspect regarding how many times any given method was called, and which arguments were used. This is particularly important to check if the code you’re testing is properly invoking the mocked class.
+
+    test_something_that_uses_my_mocked_class();
+    is(
+        $mock_meta->sub_tracking->{some_method}->@*,
+        1,
+        'some_method() called only once!'
+    );
+
+    like(
+        $mock_meta->sub_tracking->{some_method}[0]{args},
+        [ qr(), ‘arg1’, qr(arg2) ],
+        'testing call arguments for method “some_method”'
+    );
+
+    $mock_meta->clear_sub_tracking(); # so we can go again in isolation
+
+
+Note that to get the mock metadata from a given variable holding an object, you can also do:
+
+        my ($mock_meta) = mocked $actual_obj;
+
+
+=head2 Mocking the Christmas Wishlist Database
+
+For a real world example, let's see how the elves use Test2 Suite's
+mocking features to test our gift delivery system. The code looks
+roughly like this:
+
+    sub deliver_gift ($child_name, $gift_name) {
+        my $schema = GiftDB::Schema->get_active_connection();
+        my $wish = $schema->resultset('Wishlist')->single({
+            name => $child_name, gift => $gift_name
+        });
+
+        if ($wish && $wish->update({ delivered => true })) {
+            return 'delivered!';
+        }
+        return 'oh, no...';
+    }
+
+The function looks simple enough: we get an active database connection,
+select the wishlist table/resultset, and see if we can find the
+child's wishlist entry. If we have it, we update the entry to 'delivered'.
+
+But how do we test it without actually using a database? We mock the
+database, that's how!
+
+    use builtin qw( true false weaken );
+    use Test2::V0;
+
+    # mock simple resultsets and results:
+    my $mock_wishlist_rs = mock 'obj' => (
```

Comment:
Did you intend to define the 'obj' class here? Or were you trying to just make a temporary object?

Reply:
Just a temp object. This way I can point a class mock to it and use the counting features. I couldn't find a simple way to traverse and reach a mocked class's "sub-mocks", and if I return a mock object directly from a mocked class' method, it creates a new one every time so I lose the tracking. By first creating the inner mock object and then defining the mock class with a method that returns said object, I get to test if whatever is touching that part is doing so properly.